switch caching to use Postgres from Redis

JamesKunstle opened the pull request:
switches all queries over to paginate results and store in Postgres tables

implements cache_facade helper functions

adds db_init.py startup script- runs once at boot-time of compose

adds postgres container to docker compose.

moves assignment query preprocessing out of query function and into 8Knot/pages/utils/preprocessing_utils.py.

JamesKunstle commented:
Still need to:
(1) add cntrb truncation to queries
(2) make sure all time stamps are < now.
(3) update documentation for use of Postgres
(4) rebase recent changes into Postgres regime.

Creating this PR for process visibility early on.

JamesKunstle commented:
@cdolfi This is obviously a gigantic PR, so here's a walk-through guide to how I'd imagine it'd be most convenient to review everything. I'm listing the big changes so you can check them out in order before looking at the mostly-the-same changes to query / analysis worker code.

1. We're adding a "postgresql" container into the compose. This updates 'docker-compose.yaml.' We also add a 'postgres-cache' volume that gets mounted to the "postgresql" container so it has some dedicated disk space to store data.
2. To set up the postgres db for our application, I wrote a simple script in "8knot/cache-manager/db_init.py" that runs when the compose is booted. It tries to create a database called "augur_cache" in postgres and then create the tables in the cache that we need. If everything goes well it exits with a 'success' code (0) and doesn't restart. Otherwise (exit code 1) it restarts.
3. Postgres is configured in 'postgres.conf'. All write-ahead-logging is turned off, as are most data durability settings, because we can regenerate the data from Augur without any problems. We're optimizing for read/write speed here.
4. All database interactions that the application has to do (query and analysis workers) is handled by the facade written in "8Knot/cache_manager/cache_facade.py".
5. All queries are executed by logic in the cache_facade. Only the query SQL has to live in the query functions. The '%s' wildcard is where the repolists and any other arguments are passed in.
6. Because we're downloading data from Postgres and immediately storing it in another Postgres instance, all query pre-processing is moved directly into the query, such as shortening IDs to the first 15 characters and checking that the date isn't in the future.
7. All background callbacks also use the cache_facade logic. If data isn't available, they wait for it to become available by repeatedly checking the "cache_bookkeeping" table, which has rows of (cache_function, repo_id, cached_at_timestamp). 

There are still things to work on after this, like:
+ incorporating recently merged visualizations and switching queries over
+ documenting the entire Postgres flow (to be done after this is merged in case things change)
+ improving record retrieval from Augur by adding query retrieval parallelism.

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,22 +1,20 @@
 """
 TODO:
 (1) update QUERY_NAME
 (2) update 'NAME_query' found in function definition and in the function call that sets the 'ack' variable below.
```
ack is no longer used update this to current structure

JamesKunstle replied:
Ah gotcha, will change

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,22 +1,20 @@
 (5) reset df index if #4 is performed via "df = df.reset_index(drop=True)"
 (6) go to index/index_callbacks.py and import the NAME_query as a unqiue acronym and add it to the QUERIES list
-(7) delete this list when completed
+(8) specify the correct table name in the call to 'caching_wrapper'
```
also is there any reason why table name shouldnt just be the QUERY_NAME

JamesKunstle replied:
> Change "table name" to be a variable at the top to be assigned there and use the variable. Also what is "table name? Not sure what it is to be able to know what to assign to it

This answers both comments- 'table name' is the table in the database, and many queries use the same tables. There's the "commits, issues, prs, etc." tables that are all used by one or more queries.

cdolfi replied:
Im still a little confused because many queries use multiple tables

JamesKunstle replied:
Ah I can definitely be more precise-

In 8Knot, the existing tables for caching are "commits, issues, prs, contributors, issue_assignments, pr_assignments" etc. "tablename" here is what table the query executed against Augur should ultimately write to. The response from augur uses multiple tables, such as for issue_assignments, but only writes to one table in cache. 

As for why query_name, func_name, and table_name can't all be the same value:

query_name and func_name could, and likely should, be the same. I'll update that.

table_name (the table in cache) isn't necessarily the same as query_name (the function that initiates the caching flow)- it could be, but the DB schema uses shorter names for tables, like "commits" vs. "commits_query" so it seemed more flexible to just allow the developer to specify the target table rather than commit them to writing to tables that have the same name as the calling function.

cdolfi replied:
Personally, I would side to keep table name under the hood/not a part of the query creation and visualization process and stick with having the set of queries to use and get imported an used in visualizations. From my experience and talking with sean about others using 8knot the simplicity of accessing the data was a huge plus. It adds a lot of complexity

JamesKunstle replied:
okay removed this in the last update

cdolfi commented on 8Knot/pages/contributors/visualizations/contribs_by_action.py:
```diff
@@ -154,16 +166,31 @@ def toggle_popover(n, is_open):
+        ],
+    )
+
+    df = preproc_utils.contributors_df_action_naming(df)
```
I see what you are doing here and need to figure out a good way to document this on the query and visualization side for devs know that it exists, how to use it, and how to add pandas query preprocessing there

JamesKunstle replied:
++ good point

cdolfi commented on 8Knot/pages/contributors/visualizations/contribs_by_action.py:
```diff
@@ -154,16 +166,31 @@ def toggle_popover(n, is_open):
+        VIZ_ID,
+        tablename="contributors",
+        repolist=repolist,
+        column_override=[
```
What is column_override? How and why should it be used?

JamesKunstle replied:
okay removed column_override. if we want it later we can make it an optional argument

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,22 +1,20 @@
-(7) delete this list when completed
+(8) specify the correct table name in the call to 'caching_wrapper'
+(9) create a table in 8Knot/8Knot/cache_manager/db_init.py for the new data you're requesting.
+(10) delete this list when completed
```
if preprocessing_utils should be where any pandas preprocessing should be add to the list explaining what that is and how it should be used. Then any comments of pandas preprocessing of queries should be removed from the query files and moved over there for people to refer to

JamesKunstle replied:
++

cdolfi commented on 8Knot/queries/prs_query.py:
```diff
@@ -37,76 +33,51 @@ def prs_query(self, repos):
                     FROM
-                        repo r,
-                        pull_requests pr
+                        augur_data.repo r,
```
why the addition of augur_data to all queries?

JamesKunstle replied:
Because I didn't know how to set the desired schema in the psycopg2 driver configuration. I'll work on that to remove the direct references to augur_data.

JamesKunstle replied:
added the schema as a search_path parameter in the psycopg2.connect call, so we don't need 'augur_data' in the queries anymore

cdolfi commented on 8Knot/cache_manager/cache_facade.py:
```diff
@@ -0,0 +1,236 @@
+        logging.warning(f"{query_name}_DATA_QUERY - EXECUTING CACHING QUERY")
+        cache_query_results(
+            # TODO: should get this from the environment
+            db_connection_string="dbname=padres user=cali password=!baseball21 host=chaoss.tv port=5432",
```
yeah lets change this asap I dont want to have any passwords or anything in git logs

JamesKunstle replied:
oop

cdolfi reviewed (changes requested):
Ill probably go through this a couple more times throughout the week to look at things with fresh eyes. My POV for my first round(s) of review is going to be from the "data scientist" / "visualization maker" persona of things being compartmentalized where understanding of the caching system is not necessary, just enough to understand how to get their query data and make a new query if necessary. Im going to purposely only look at the queries and visualizations for now until we get to a place that we like that side then ill review the more dense caching code. 

On the documentation side: we might want to merge this into a different branch on 8knot and get all of the documentation up to date first before going to dev

Also, how is the speed of the edited queries? Many times the choices I made of doing things in pandas vs sql was from efficiency

JamesKunstle commented:
> Ill probably go through this a couple more times throughout the week to look at things with fresh eyes. My POV for my first round(s) of review is going to be from the "data scientist" / "visualization maker" persona of things being compartmentalized where understanding of the caching system is not necessary, just enough to understand how to get their query data and make a new query if necessary. Im going to purposely only look at the queries and visualizations for now until we get to a place that we like that side then ill review the more dense caching code.
> 
> On the documentation side: we might want to merge this into a different branch on 8knot and get all of the documentation up to date first before going to dev
> 
> Also, how is the speed of the edited queries? Many times the choices I made of doing things in pandas vs sql was from efficiency

(1) Yeah this PR is targeting a new branch. We'll want to get everything from the last couple of weeks nicely merged into this branch before we swap it out into dev.

(2) The edited queries aren't noticeably slower with the additional filtering and transformation steps. They're applied "at projection time" rather than "at query time" in DBMS terms, so they're relatively fast. This new regime sorta requires that we do things in SQL rather than Pandas when we're grabbing the data from Augur, because we're no longer using df's as an intermediate data store- we're transferring records directly from Postgres into Postgres.

JamesKunstle commented on 8Knot/queries/query_template.py:
```diff
@@ -1,22 +1,20 @@
 (2) update 'NAME_query' found in function definition and in the function call that sets the 'ack' variable below.
 'NAME' should be the same as QUERY_NAME
-(3) paste SQL query in the query_string
+(3) paste SQL query in the query_string. Make sure it's wild-carded for psycopg2
```
Okay totally, that makes sense.

JamesKunstle commented:
@cdolfi I responded to the comments you made- will make requested changes later.

sgoggins commented on 8Knot/queries/company_query.py:
```diff
@@ -42,82 +37,51 @@ def company_query(self, repos):
                     SELECT
-                        c.cntrb_id,
-                        c.created_at AS created,
+                        left(c.cntrb_id::text, 15), -- first 15 characters of the uuid
```
This is kind of a bad idea because those first 15 can be non-unique.

cdolfi replied:
When looking at all of our cntrb_ids the most unique numbers before the 0s start was 8

sgoggins commented on 8Knot/queries/company_query.py:
```diff
@@ -42,82 +37,51 @@ def company_query(self, repos):
-                        c.cntrb_id,
-                        c.created_at AS created,
+                        left(c.cntrb_id::text, 15), -- first 15 characters of the uuid
+                        timezone('utc', c.created_at) AS created,
```
This is good.

JamesKunstle replied:
do these projection-time modifications add significant overhead?

JamesKunstle replied:
++ awesome that's what I was hoping

sgoggins commented on 8Knot/queries/contributors_query.py:
```diff
@@ -42,85 +37,44 @@ def contributors_query(self, repos):
-                        rank
+                        ca.repo_id as id,
+                        ca.repo_name,
+                        left(ca.cntrb_id::text, 15) as cntrb_id, -- first 15 characters of the uuid
```
still a bad idea.

sgoggins reviewed (commented):
@JamesKunstle : I started reviewing this and I have a bucket of questions I think are more conceptual, and would benefit from conversation.

JamesKunstle commented:
@sgoggins could you look at how I configure the postgres database in postgres.conf? I'm configuring based on the container being scheduled with 4GB

JamesKunstle commented on 8Knot/cache_manager/db_init.py:
```diff
@@ -0,0 +1,231 @@
+    conn = pg.connect(
+        dbname="postgres",
+        user="postgres",
+        password="password",
```
does it matter whether the password is secure or not if the connected-to instance is only available via internal networking? this db instance lives on OCP or in the compose so it should only be accessible by the app server as the single client.

JamesKunstle replied:
moved this into a common file for connection strings so we can set everything up with environment variables in the future.

JamesKunstle replied:
okay that's totally reasonable rationale. I'll update it so that it uses a password from the environment!

JamesKunstle commented on 8Knot/pages/affiliation/visualizations/commit_domains.py:
```diff
@@ -134,15 +133,28 @@ def toggle_popover(n, is_open):
-        df = cache.grabm(func=cq, repos=repolist)
+    while not_cached := cf.get_uncached(func_name=cmq.__name__, repolist=repolist):
+        logging.warning(f"COMMITS_OVER_TIME_VIZ - WAITING ON DATA TO BECOME AVAILABLE")
+        time.sleep(0.5)
```
definitely- next feature removes the blind polling entirely, switching to waiting on the collecting process via the Celery tasks lifecycle API.

JamesKunstle commented on docker-compose.yml:
```diff
@@ -81,6 +113,21 @@ services:
     restart: always
 
+  postgres-cache:
+    image: docker.io/library/postgres:latest
```
ah okay that's a good point

JamesKunstle replied:
should I just pick 15 or 16 or some other major version?

JamesKunstle replied:
Okay! That totally makes sense.

JamesKunstle commented on 8Knot/queries/pr_assignee_query.py:
```diff
@@ -40,65 +35,45 @@ def pr_assignee_query(self, repos):
+                        pa.id in %s
+                        and pa.created < now()
+                        and pa.closed < now()
+                        and pa.assign_date < now()
```
Yeah the tiny corner-cases are acceptable. The major problem is wanting to avoid retrieving data that are mis-labelled as having occurred in the future

JamesKunstle commented on 8Knot/queries/prs_query.py:
```diff
@@ -37,76 +33,51 @@ def prs_query(self, repos):
 
+    # reformat cntrb_id
+    df["cntrb_id"] = df["cntrb_id"].astype(str)
+    df["cntrb_id"] = df["cntrb_id"].str[:15]
```
we should probably NOT be truncating user IDs in the first place at the data-layer, but yes, I agree with your point. I'll think of how to parameterize this.

JamesKunstle commented on docker-compose.yml:
```diff
@@ -7,21 +19,32 @@ services:
+        "1",
+        "--threads",
+        "2",
+      ]
```
yes you're right- I believe this was an artifact of my VSCode linter not agreeing with the Black formatter that runs w/ pre-commit.

JamesKunstle commented:
> The PR kinda show there is a huge potentail to refactor around the methods that use caching (exemple create_top_k_cntrbs_graph ). They are all the same strucutre minus 1 line change.

Yeah there are, most certainly. We made an early design choice to optimize for functional clarity at the visualization level, which has lead to a lot of repeating-of-ourselves at the scale we're operating at now. I have some notes on how we can refine this in the future.

JamesKunstle commented:
@cdolfi I think I addressed all of your review comments

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,26 +1,30 @@
+    where you intend to use the repos list. The '%s' will be replaced with a literal representation of the 'repos' variable.
+(3) Navigate to 8Knot/pages/index/index_callbacks.py. Import your new query as a unique acronym. add it to the QUERIES list.
+    this registers your query to be scheduled when a user requests new data.
+(4) Create a table in 8Knot/8Knot/cache_manager/db_init.py for the new data you're retrieving from augur. Name the table identically to
```
Need a little more details here on the steps on what it means to "create a table". Either template/ step by step explanation in here or db_init on what to do

JamesKunstle replied:
Okay good point

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,26 +1,30 @@
+    (a) be done in the SQL itself, so all records retrieved from Augur are usable or
+
+    (b) become analysis pre-processing steps that run on-the-fly on user request. Utilities for this kind of
+        processing are typically in 8Knot/8Knot/pages/utils/preprocessing_utils.py, but YMMV.
```
What does YMMV mean?

cdolfi replied:
^ YMMV still there

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,26 +1,30 @@
-(5) reset df index if #4 is performed via "df = df.reset_index(drop=True)"
-(6) go to index/index_callbacks.py and import the NAME_query as a unqiue acronym and add it to the QUERIES list
-(7) delete this list when completed
+(1) Rename the query function to something informative. Replace other instances of "NAME_query" with your
```
Is there a way to make "NAME_query" a single variable at the top so it only needs to be done in one place?

JamesKunstle replied:
uhhh I think so? I'll check it out. that'd be naming the function via f-strings I think. I'm not sure if that's supported.

JamesKunstle replied:
No, we can't define the name of a function as a variable and then use it to define the function. Having the user ctrl-f to find all instances of the 'query_NAME' thing should be fine though. I'll add that to the template notes.

cdolfi commented on 8Knot/queries/query_template.py:
```diff
@@ -1,26 +1,30 @@
+
+        If you choose this path, PLEASE DOCUMENT this behavior.
 """
 
```
All current visualizations are still using "QUERY_NAME" . We need to decide if we are still going to use that structure and keep that in the template or change all of the queries to be in consistent structure

JamesKunstle replied:
I'm seeing that all the queries use that variable for custom-naming the logs that are produced during collection. In the visualization template, that variable is just used to note that we're importing a query by its name. We can either remove the use of QUERY_NAME in the queries, or we can leave it as. I don't really mind either, because we're not using it in any APIs downstream.

cdolfi replied:
Yeah lets change it for it is consistent. Mainly for people to have a reference when making new queries and might get confused from the template vs the queries that exist.

cdolfi reviewed (changes requested):
A few additional questions and comments but from the viz/query creation standpoint this is getting pretty close. Once we smooth out these edges Ill take a step back on the review process and once youve worked through misc and seans comments lmk and Ill do another full review

jberkus commented:
So, I feel like I'm missing some big picture information here.

1. Why drop Redis, which is generally used as an effective caching database?  What's the goal?
2. Why not use materialized views?
3. Why put the cached data in a different DB on the same server?  Or is that just a prelude to moving it to a db on a separate server from the main server?

Thanks, with that framing I'll be able to supply more useful advice.

JamesKunstle commented:
> So, I feel like I'm missing some big picture information here.
> 
> 1. Why drop Redis, which is generally used as an effective caching database?  What's the goal?
> 2. Why not use materialized views?
> 3. Why put the cached data in a different DB on the same server?  Or is that just a prelude to moving it to a db on a separate server from the main server?
> 
> Thanks, with that framing I'll be able to supply more useful advice.

Very good questions.

1. Redis is a great caching database, we were just using it in a way that wasn't going to scale well, and that required a lot more memory than we had. Our data is bigger than memory, so Redis no longer works for us. Our data is also table-like when cached, so it made little sense to force Redis or another k/v db to be a row-oriented database (mget-ing loads of values by keys) when something relational is already very good at that, and supports retrieval optimization stuff like indexing. I looked into a variety of nosql databases as well and concluded that, of the available options (Cockroach, Maria, Mongo, etc.) Postgres was still probably the best bet. If there was an even simpler SQL db on the market that was as well-vetted as pg I would have considered it, but Postgres fit our needs well. (I did try sqlite and duckdb, and they were cool, but insufficient on OCP.)

2. We can, and do, use materialized views as much as possible. Caching is a matter of data locality. The main Augur database that we use is in Columbia, Missouri, and I believe that our OCP boxes are in US-west. In testing, it's orders of magnitude faster to query data once from Augur, store it in a db that lives on the same OCP boxes as our app server, and serve the data hence from there.

3. answered in 2

I hope that answers your questions well. 

My ideal roadmap involves having the main database and the application servers live on the same hardware so we can gut this data caching layer and focus on service-time optimizations. We're not there yet, so this is a medium-term fix that gives us a lot more time to focus on analysis rather than data engineering.

JamesKunstle commented:
@cdolfi
- Updated db_init.py to have more instructions,
- Changed "ymmv" to "your mileage may vary"

@mscherer 
- password for Postgres cache set in environment

@sgoggins 
- what do you think about the cntrb_id being shortened based on what we said? We could keep them full-length for analysis and shorten them for visualization too, which would solve both the aesthetic and the correctness arguments.


Once all reviewers are satisfied, we can merge this PR into the temp branch that it's pointed at and start the process of rebasing the changes that have been added to the 'dev' branch while I've been working on this feature.

After that, I'll add the enterprisedb image into the Openshift definition of the app and write the cache invalidation cronjob that cleans out the cache once every 48 hours.

cdolfi commented on 8Knot/cache_manager/db_init.py:
```diff
@@ -0,0 +1,256 @@
+creation block from those below. Given that you're creating a table
+for a query in the 'queries/' folder, name the table the same name
+as the query function. Name the columns of the table, and give their
+types, and everything should work!
```
Is there a resource that can list the type options and what they are? Not specific to 8knot, but a general resource if that makes sense?

cdolfi commented on 8Knot/queries/commits_query.py:
```diff
@@ -1,16 +1,11 @@
 import os
 
-QUERY_NAME = "COMMITS"
+QUERY_NAME = "commits_query"
```
For all query files: update to new format not using QUERY_NAME

JamesKunstle replied:
got it, switched everything over.

cdolfi reviewed (changes requested):
Only a few more small changes on my side! Would be great for @sgoggins to take a look at these changes from a viz making perspective as he has been helping students build visualizations and might have additional feedback

JamesKunstle commented on 8Knot/cache_manager/db_init.py:
```diff
@@ -0,0 +1,265 @@
+
+        return 0
+
+    except Exception as e:
```
That's definitely fair. In this case, I specifically want ANYTHING that bubbles up to caught and trigger a script failure. I don't have any error handling that is exception-specific, I just need a log report of which Exception specifically got thrown. But for other situations I 100% agree that we'd want finer-grained exception handling.
